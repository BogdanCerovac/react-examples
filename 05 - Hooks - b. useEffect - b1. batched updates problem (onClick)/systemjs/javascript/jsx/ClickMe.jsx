/*! Mozilla Public License Version 2.0 !*/
/*! Copyright Â© 2018 Rick Beerendonk   !*/

/* eslint react/prop-types:"off" */

import React from 'react';

// SystemJS limitation (we should be able to do this directly in import)
const { useState } = React;

function ClickMe() {
  const [count, setCount] = useState(0);

  return (
    <a
      onClick={() => {
        setCount(count + 0.5);
        setCount(count + 0.5);
      }}
    >
      {`This link has been clicked ${count} times`}
    </a>
  );
}

export default ClickMe;
