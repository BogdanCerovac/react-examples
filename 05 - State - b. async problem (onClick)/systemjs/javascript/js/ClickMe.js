/*! Mozilla Public License Version 2.0 !*/
/*! Copyright Â© 2018 Rick Beerendonk   !*/

/* eslint react/prop-types:"off" */

import React from 'react';

export default class ClickMe extends React.Component {
  constructor(props) {
    super(props);
    this.state = { count: 0 };

    // Bind all non-react methods to this.
    this.onClick = this.onClick.bind(this);
  }
  onClick() {
    // Updates are batched for performance reasons.
    // Therfor this won't work:
    this.setState({ count: this.state.count + 0.5 });
    this.setState({ count: this.state.count + 0.5 });
  }
  render() {
    return React.createElement('a', { onClick: this.onClick },
      'This link has been clicked ',
      this.state.count,
      ' times'
    );
  }
}
