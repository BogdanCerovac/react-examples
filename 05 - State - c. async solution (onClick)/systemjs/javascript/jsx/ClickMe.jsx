/*! Mozilla Public License Version 2.0 !*/
/*! Copyright Â© 2018 Rick Beerendonk   !*/

/* eslint react/prop-types:"off" */

import React from 'react';

export default class ClickMe extends React.Component {
  constructor(props) {
    super(props);
    this.state = { count: 0 };

    // Bind all non-react methods to this.
    this.onClick = this.onClick.bind(this);
  }
  onClick() {
    // this.state may be updated asynchronously:
    this.setState(prevState => ({ count: prevState.count + 1 }));
  }
  render() {
    return (
      <a onClick={this.onClick}>
        {`This link has been clicked ${this.state.count} times`}
      </a>
    );
  }
}
