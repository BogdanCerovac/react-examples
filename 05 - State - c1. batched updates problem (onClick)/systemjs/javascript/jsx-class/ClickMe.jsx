/*! Mozilla Public License Version 2.0 !*/
/*! Copyright Â© 2018 Rick Beerendonk   !*/

import React from 'react';

export default class ClickMe extends React.Component {
  constructor(props) {
    super(props);
    this.state = { count: 0 };

    // Bind all non-react methods to this.
    this.onClick = this.onClick.bind(this);
  }
  onClick() {
    // Updates are batched for performance reasons.
    // Therefore this won't work:
    this.setState({ count: this.state.count + 0.4 });
    this.setState({ count: this.state.count + 0.6 });
  }
  render() {
    return (
      <a onClick={this.onClick}>
        {`This link has been clicked ${this.state.count} times`}
      </a>
    );
  }
}
