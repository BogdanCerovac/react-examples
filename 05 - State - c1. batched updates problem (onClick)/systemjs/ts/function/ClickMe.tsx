/*! European Union Public License version 1.2 !*/
/*! Copyright Â© 2018 Rick Beerendonk          !*/

import React from 'react';

function ClickMe(): JSX.Element {
  const [count, setCount]: [
    number,
    React.Dispatch<React.SetStateAction<number>>
  ] = React.useState(0);

  const handleClick = () => {
    // Updates are batched for performance reasons.
    // Therefore this won't work:
    setCount(count + 0.4);
    setCount(count + 0.6);
  };

  return (
    <a onClick={handleClick}>{`This link has been clicked ${count} times`}</a>
  );
}

export default ClickMe;
