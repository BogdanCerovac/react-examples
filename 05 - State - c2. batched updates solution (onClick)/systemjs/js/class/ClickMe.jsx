/*! Mozilla Public License Version 2.0 !*/
/*! Copyright Â© 2018 Rick Beerendonk   !*/

import React from 'react';

export default class ClickMe extends React.Component {
  constructor(props) {
    super(props);
    this.state = { count: 0 };

    // Bind all non-react methods to this.
    this.handleClick = this.handleClick.bind(this);
  }
  handleClick() {
    // Updates are batched for performance reasons.
    // Therefore this should be used:
    this.setState(state => ({ count: state.count + 0.4 }));
    this.setState(state => ({ count: state.count + 0.6 }));
  }
  render() {
    return (
      <a onClick={this.handleClick}>
        {`This link has been clicked ${this.state.count} times`}
      </a>
    );
  }
}
