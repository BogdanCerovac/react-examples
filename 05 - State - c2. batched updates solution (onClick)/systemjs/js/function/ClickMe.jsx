/*! European Union Public License version 1.2 !*/
/*! Copyright Â© 2018 Rick Beerendonk          !*/

import React from 'react';

function ClickMe() {
  const [count, setCount] = React.useState(0);

  return (
    <a
      onClick={() => {
        setCount(c => c + 0.4);
        setCount(c => c + 0.6);
      }}
    >
      This link has been clicked {count} times
    </a>
  );
}

export default ClickMe;
