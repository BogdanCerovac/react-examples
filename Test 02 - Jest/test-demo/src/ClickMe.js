import React from 'react';

class ClickMe extends React.Component {
  constructor(props) {
    super(props);
    this.state = { count: 0 };

    // Bind all non-react methods to this.
    this.onClick = this.onClick.bind(this);
  }
  onClick() {
    // this.state may be updated asynchronously:
    this.setState(state => ({ count: state.count + 1 }));
  }
  render() {
    return (
      <a onClick={this.onClick}>
        {`This link has been clicked ${this.state.count} times`}
      </a>
    );
  }
}

export default ClickMe;
