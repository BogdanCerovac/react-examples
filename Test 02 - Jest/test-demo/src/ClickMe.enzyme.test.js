import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ClickMe from './ClickMe';

/* eslint-env jest */

/// Enzyme ///

it('renders 0 snapshot (enzyme)', () => {
  const wrapper = shallow(<ClickMe />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

it('renders 1 after click snapshot (enzyme)', () => {
  const wrapper = shallow(<ClickMe />);
  wrapper.find('a').simulate('click');
  expect(toJson(wrapper)).toMatchSnapshot();
});
