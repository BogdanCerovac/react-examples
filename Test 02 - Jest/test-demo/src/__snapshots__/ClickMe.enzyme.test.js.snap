// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 0 snapshot (enzyme) 1`] = `
<a
  onClick={[Function]}
>
  This link has been clicked 0 times
</a>
`;

exports[`renders 1 after click snapshot (enzyme) 1`] = `
<a
  onClick={[Function]}
>
  This link has been clicked 1 times
</a>
`;
